feat(rpc): support side chains in getrawtransaction

conradoplg opened the pull request:
## Motivation

Closes #9873

## Solution

Add new state methods to find blocks and transactions in any chains.
Change `getrawtransaction` to use those.

I also changed the `invalidateblock` and `reconsiderblock`, which are used in the test, because they weren't working in JSON-RPC.

### Tests

I added a Python test for this, since it seemed easier. I think those aren't being run in CI yet, but we can do that later.

```
PYTHON_DEBUG=1 CARGO_BIN_EXE_zebrad=~/zebra/target/debug/zebrad PYTHON_DEBUG=1 ./qa/rpc-tests/getrawtransaction_sidechain.py
```

### Specifications & References

<!-- Provide any relevant references. -->

### Follow-up Work

<!--
- If there's anything missing from the solution, describe it here.
- List any follow-up issues or PRs.
- If this PR blocks or depends on other issues or PRs, enumerate them here.
-->

### PR Checklist

<!-- Check as many boxes as possible. -->

- [ ] The PR name is suitable for the release notes.
- [ ] The solution is tested.
- [ ] The documentation is up to date.

conradoplg committed:
rpc: support side chains in getrawtransaction

conradoplg committed:
add test

conradoplg committed:
fix existing test

upbqdn commented on zebra-rpc/src/methods/types/transaction.rs:
```diff
@@ -158,13 +158,14 @@ pub struct TransactionObject {
     pub(crate) hex: SerializedTransaction,
-    /// The height of the block in the best chain that contains the tx or `None` if the tx is in
-    /// the mempool.
+    /// The height of the block in the best chain that contains the tx, -1 is
```
```suggestion
    /// The height of the block in the best chain that contains the tx, -1 if
```

upbqdn commented on zebra-rpc/src/methods.rs:
```diff
@@ -637,15 +639,15 @@ pub trait Rpc {
     // TODO: Invalidate block hashes even if they're not present in the non-finalized state (#9553).
     #[method(name = "invalidateblock")]
-    async fn invalidate_block(&self, block_hash: block::Hash) -> Result<()>;
+    async fn invalidate_block(&self, block_hash: String) -> Result<()>;
```
Was it the error code that was problematic?

conradoplg replied:
`block::Hash` does not deserialize from a hex string by default. It was being deserialized from an integer array. (We could change the `block::Hash` serde serialization format but that might break something else... We can probably review all serde implementations in https://github.com/ZcashFoundation/zebra/issues/9892)

conradoplg committed:
Update zebra-rpc/src/methods/types/transaction.rs

Co-authored-by: Marek <mail@marek.onl>

upbqdn commented on zebra-rpc/src/methods/types/transaction.rs:
```diff
@@ -625,8 +626,16 @@ impl TransactionObject {
         let block_time = block_time.map(|bt| bt.timestamp());
         Self {
             hex: tx.clone().into(),
-            height: height.map(|height| height.0),
-            confirmations,
+            height: if in_active_chain.unwrap_or_default() {
+                height.map(|height| height.0 as i32)
+            } else {
+                Some(-1)
+            },
+            confirmations: if in_active_chain.unwrap_or_default() {
+                confirmations
+            } else {
+                Some(0)
+            },
```
We set `in_active_chain` to `Some(false)` here: https://github.com/zcashfoundation/zebra/blob/97a805874bb6c61d688ccbb1709ba9bf10b899b8/zebra-rpc/src/methods.rs#L1757 when the tx comes from the mempool, so this logic returns `Some(-1)` for `height` and `Some(0)` for `confirmations`, but the updated docs above say that the values should be `None` and `None`.

upbqdn replied:
I updated the link in the comment above because I noticed the original one didn't point to the exact line I intended to reference.

conradoplg replied:
Thanks, nice catch. This led me into a big rabbit hole that made me realize the test wasn't working properly. I addressed this and the other stuff in 58fc1a2d2

oxarbitrage commented on zebra-rpc/qa/rpc-tests/getrawtransaction_sidechain.py:
```diff
@@ -0,0 +1,61 @@
+    nuparams,
+    start_nodes,
+    wait_and_assert_operationid_status,
+)
```
```suggestion
from test_framework.util import start_nodes
```

oxarbitrage replied:
Lint: From util imports, the test just use `start_nodes`.

oxarbitrage commented on zebra-rpc/qa/rpc-tests/getrawtransaction_sidechain.py:
```diff
@@ -0,0 +1,61 @@
+# file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#
+# Test the effect of reorgs on the Orchard commitment tree.
```
```suggestion
# Test getrawtransaction on side chains
```

oxarbitrage commented on zebra-rpc/qa/rpc-tests/getrawtransaction_sidechain.py:
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022-2024 The Zcash developers
```
```suggestion
# Copyright (c) 2025 The Zcash developers
```

Optional: I am not sure how this is going to be in the future, but for a new file not ported from zcashd what i had been  doing is: https://github.com/ZcashFoundation/zebra/blob/main/zebra-rpc/qa/rpc-tests/feature_backup_non_finalized_state.py#L2

oxarbitrage reviewed (commented):
I added some comments that i noticed in the python framework only.

conradoplg committed:
fix RPC return issue; and multiple other issues

- Change AnyTx to include the block of tx

- Use random coinbase data in generate() to make txs unique

- Propagate submit_block() errors in generate()

- Allow comitting blocks even if there was an invalidated block with the same height

conradoplg committed:
Apply suggestions from code review

Co-authored-by: Alfredo Garcia <oxarbitrage@gmail.com>

conradoplg reviewed (commented):
While addressing @upbqdn 's comment I found multiple issues which were all fixed in 58fc1a2d2

- The `generate()` after the `reconsiderblock()` was failing but an error wasn't being returned, so the test kept going and with the way it was written, it passed but it wasn't actually looking into blocks across forks. I changed `generate()` to propagate any error from `submit_block()`
- The reason that was failing was a check that would prevent committing a block with a height of a previously invalidated block. That doesn't seem to make sense to me, so I changed the check to return an error only if the hash is the same of a previously invalidated block
- After fixing that, there was another issue: after `invalidateblock()`, `generate()` would generate a block exactly identical to the invalidated one, which would then fail to commit. To fix that, I changed `generate()` to use random coinbase data for each block, which makes the coinbase tx unique and its block too.

conradoplg committed:
Merge branch 'main' of https://github.com/ZcashFoundation/zebra into getrawtransaction-sidechain

conradoplg commented:
@mergifyio update

mergify[bot] committed:
Merge branch 'main' into getrawtransaction-sidechain

upbqdn commented on zebra-rpc/qa/rpc-tests/getrawtransaction_sidechain.py:
```diff
@@ -0,0 +1,71 @@
+        hashes = n.generate(2)
+
+        # Get a transaction from block at height 2
+        block_hash_a = n.getbestblockhash()
+        txid_a = n.getblock("2", 1)['tx'][0]
```
It might be worth either obtaining the block and its hash from a single call or asserting that the hash corresponds to the fetched block.

conradoplg replied:
I added a check in 7e5c15e16 which should achieve the same thing

upbqdn commented on zebra-rpc/src/methods/types/get_block_template.rs:
```diff
@@ -519,6 +518,17 @@ where
         self.extra_coinbase_data.clone()
     }
 
+    /// Changes the extra coinbase data.
```
```suggestion
    /// Changes the extra coinbase data.
    ///
    /// # Panics
    ///
    /// If `extra_coinbase_data` exceeds [`EXTRA_COINBASE_DATA_LIMIT`].
```

upbqdn commented on zebra-state/src/request.rs:
```diff
@@ -733,6 +733,14 @@ pub enum Request {
+    /// Returns
+    ///
+    /// * [`Response::AnyChainTransaction(Some(Arc<Transaction>))`](Response::AnyChainTransaction) if the transaction is in the best chain;
+    /// * [`Response::AnyChainTransaction(None)`](Response::AnyChainTransaction) otherwise.
```
```suggestion
    /// Looks up a transaction by hash in any chain.
    ///
    /// Returns
    ///
    /// * [`Response::AnyChainTransaction(Some(AnyTx))`](Response::AnyChainTransaction)
    ///   if the transaction is in any chain;
    /// * [`Response::AnyChainTransaction(None)`](Response::AnyChainTransaction)
    ///   otherwise.
```

upbqdn commented on zebra-state/src/service.rs:
```diff
@@ -1579,6 +1604,37 @@ impl Service<ReadRequest> for ReadStateService {
                 .wait_for_panics()
             }
 
+            // Used by the getblock (verbose) RPC.
```
```suggestion
```

upbqdn commented on zebra-rpc/src/methods.rs:
```diff
@@ -1812,40 +1817,63 @@ where
+                        {
+                            zebra_state::ReadResponse::BlockHash(block_hash) => block_hash,
+                            _ => unreachable!(
+                                "unmatched response to a `AnyChainTransaction` request"
```
```suggestion
                                "unmatched response to a `BestChainBlockHash` request"
```

upbqdn commented on zebra-rpc/src/methods.rs:
```diff
@@ -1812,40 +1817,63 @@ where
+                let hex = match tx {
+                    AnyTx::Mined(mined_tx) => mined_tx.tx.into(),
+                    AnyTx::Side((transaction, _block_hash)) => transaction.into(),
+                };
```
We could get rid of this match if we impl'd `From` for `AnyTx`.

conradoplg replied:
Done in 7e5c15e16

upbqdn commented on zebra-rpc/src/methods.rs:
```diff
@@ -2960,9 +3005,20 @@ where
+                .submit_block(hex_proposal_block, None)
                 .await
                 .map_error(server::error::LegacyCode::default())?;
+            match r {
+                SubmitBlockResponse::Accepted => { /* pass */ }
+                SubmitBlockResponse::ErrorResponse(response) => {
+                    return Err(ErrorObject::owned(
+                        server::error::LegacyCode::Misc.into(),
+                        format!("block was rejected: {:?}", response),
+                        None::<()>,
+                    ));
+                }
+            }
```
Could we not go with https://github.com/zcashfoundation/zebra/blob/ab390bb555238c1c580971aef542e32c05ddacc2/zebra-rpc/src/server/error.rs#L109?

conradoplg replied:
Unfortunately not, `response` is not a `Result`

upbqdn commented on zebra-state/src/service/read/block.rs:
```diff
@@ -149,6 +149,47 @@ where
+    // # Correctness
+    //
+    // It is ok to do this lookup in multiple different calls. Finalized state updates
+    // can only add overlapping blocks, and hashes are unique.
```
What do overlapping blocks mean?

conradoplg replied:
(For context, I just adapted the comment from `mined_transaction`.)

This is about the issue where blocks could be committed into the state while the function is running; in that case the copies of the non-finalized chains will have a block that overlaps a committed block.

conradoplg committed:
Apply suggestions from code review

Co-authored-by: Marek <mail@marek.onl>

conradoplg committed:
address comments

upbqdn commented on zebra-rpc/src/methods/types/transaction.rs:
```suggestion
```

upbqdn commented on zebra-rpc/src/methods/types/transaction.rs:
```suggestion
```

conradoplg committed:
remove unused import
